kolena/_experimental/trace/trace.py

```diff
@@ -0,0 +1,241 @@
+        try:
+            self.existing_dataset = _load_dataset_metadata(self.dataset_name)
+            if sorted(id_fields) != sorted(self.existing_dataset.id_fields):
+                raise ValueError(f"Id Fields {id_fields} do not match existing dataset id fields")
+            self.id_fields = self.existing_dataset.id_fields
+        except NotFoundError:
+            self.existing_dataset = None
+            if not id_fields:
+                id_fields = [KOLENA_DEFAULT_ID]
+            self.id_fields = id_fields
+        for field in self.id_fields:
+            if field not in self.signature.parameters and field != KOLENA_DEFAULT_ID:
+                raise ValueError(f"Id Field {field} not found in function signature")
+        self.datapoints = []
+        self.results = defaultdict(list)
+        self.last_update = time.time()
+        self.task_ongoing = None
+        self.sync_interval = sync_interval
+        self.record_timestamp = record_timestamp
+        functools.update_wrapper(self, func)
+        atexit.register(self._clean_up)
+
+    def _add_id_fields(self, datapoint: Dict, result: Dict) -> None:
+        for field in self.id_fields:
+            if field == KOLENA_DEFAULT_ID:
+                call_id = uuid.uuid4().hex
+                datapoint[KOLENA_DEFAULT_ID] = call_id
+                result[KOLENA_DEFAULT_ID] = call_id
+            else:
+                field_value = datapoint.get(field)
+                if field_value is None:
+                    raise ValueError(f"Id Field {field} cannot be None in datapoint input")
+                result[field] = field_value
+
+    def __call__(self, *args, **kwargs):
+        bounded_arguments = self.signature.bind(*args, **kwargs)
+        bounded_arguments.apply_defaults()
+        arguments = bounded_arguments.arguments
+
+        start_time = datetime.now()
+        output = self.func(**arguments)
+        end_time = datetime.now()
+        datapoint = _serialize_item(arguments)
+        result = _serialize_item(output)
+        if not isinstance(result, dict):
+            result = {"result": result}
+        if self.record_timestamp:
+            datapoint[KOLENA_TIMESTAMP_KEY] = start_time.isoformat()
+            result[KOLENA_TIMESTAMP_KEY] = end_time.isoformat()
+            result[KOLENA_TIME_ELAPSED_KEY] = (end_time - start_time).total_seconds()
+        self._add_id_fields(datapoint, result)
+        self.datapoints.append(datapoint)
+        model_name = self.model_name
+        if self.model_name_field and arguments.get(self.model_name_field) is not None:
+            model_name = arguments.get(self.model_name_field)
+        self.results[model_name].append(result)
+        if time.time() - self.last_update > self.sync_interval and (
+            self.task_ongoing is None or not self.task_ongoing.is_alive()
+        ):
+            self.task_ongoing = threading.Thread(target=self._push_data)
+            self.task_ongoing.start()
+        return output
```

Comment:
a general question: how's the thread safeness? do we expect any new results appended when pushing the data? or we have lock to prevent it?

Reply:
Maybe concurrent write to the list and slicing the list might cause race condition, I will add a lock there. But it is ok to append new things to the list while pushing, this is safe and intentional, we don't want to slowdown client side execution